### PR TITLE
69 implement search view functionality

### PR DESCRIPTION
**Pull request by bogdan-evtushenko**

### CODE REVIEW

File: app/src/main/java/com/bogdan/codeforceswatcher/features/problems/ProblemsAdapter.kt [hunk 1]
```diff
@@ -65,11 +69,49 @@ class ProblemsAdapter(
...
-    fun setItems(problemsList: List<Problem>) {
+    fun setItems(problemsList: List<Problem>, constraint: String) {
         items = problemsList
+        showingItems = buildFilteredList(constraint)
         notifyDataSetChanged()
     }
 
+    private val problemFilter: Filter = object : Filter() {
+        override fun performFiltering(constraint: CharSequence?): FilterResults {
+            val filteredList = mutableListOf<Problem>()
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/problems/ProblemsAdapter.kt]: You can assign `if` directly.

File: app/src/main/java/com/bogdan/codeforceswatcher/features/problems/ProblemsAdapter.kt [hunk 2]
```diff
@@ -65,11 +69,49 @@ class ProblemsAdapter(
...
+            if (constraint.isNullOrEmpty()) {
+                filteredList.addAll(items)
+            } else {
+                filteredList.addAll(buildFilteredList(constraint.toString()))
+            }
+            val results = FilterResults()
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/problems/ProblemsAdapter.kt]: Can't you pass `values` in constructor? If not, you always can do `apply`.

File: app/src/main/java/com/bogdan/codeforceswatcher/features/problems/ProblemsAdapter.kt [hunk 3]
```diff
@@ -65,11 +69,49 @@ class ProblemsAdapter(
...
+
+        override fun publishResults(constraint: CharSequence, results: FilterResults) {
+            showingItems.clear()
+            showingItems.addAll(results.values as List<Problem>)
+            notifyDataSetChanged()
+        }
+    }
+
+    private fun buildFilteredList(constraint: String): MutableList<Problem> {
+        val lowerCaseConstraint = constraint.toLowerCase(Locale.getDefault())
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/problems/ProblemsAdapter.kt]: Not sure locale is needed here and in this method overall. Why can't we just `lowercase` without it?

File: app/src/main/java/com/bogdan/codeforceswatcher/features/problems/ProblemsFragment.kt [hunk 4]
```diff
@@ -30,6 +36,24 @@ class ProblemsFragment : Fragment(), StoreSubscriber<ProblemsState>, SwipeRefres
...
     }
 
+    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
+        inflater?.inflate(R.menu.menu_search, menu)
+        val searchItem = menu?.findItem(R.id.action_search)
+        searchView = searchItem?.actionView as? SearchView
+
+        searchView?.imeOptions = EditorInfo.IME_ACTION_DONE
+        searchView?.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
+            override fun onQueryTextSubmit(query: String?): Boolean {
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/problems/ProblemsFragment.kt]: One-line, please.

File: app/src/main/java/com/bogdan/codeforceswatcher/features/problems/ProblemsFragment.kt [hunk 5]
```diff
@@ -61,7 +85,8 @@ class ProblemsFragment : Fragment(), StoreSubscriber<ProblemsState>, SwipeRefres
         swipeRefreshLayout.isRefreshing = (state.status == ProblemsState.Status.PENDING)
         problemsAdapter.setItems(
             if (state.isFavourite) state.problems.filter { it.isFavourite }.sortedByDescending { it.contestTime }
-            else state.problems.sortedByDescending { it.contestTime }
+            else state.problems.sortedByDescending { it.contestTime },
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/problems/ProblemsFragment.kt]: Can't really get what's `,` doing here.

File: app/src/main/java/com/bogdan/codeforceswatcher/features/problems/redux/requests/ProblemsRequests.kt [hunk 6]
```diff
@@ -40,7 +37,10 @@ class ProblemsRequests {
                     var problems = listOf<Problem>()
                     withContext(Dispatchers.IO) {
                         problems = mergeProblems(problemsEn, problemsRu)
-                        bindProblemsToContests(problems)
+                        for (i in 0..3) {
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/problems/redux/requests/ProblemsRequests.kt]: Oh, please, don't make things like this ever 🙏Add `delay` if request isn't initiated by user, for example.

File: app/src/main/java/com/bogdan/codeforceswatcher/features/problems/redux/requests/ProblemsRequests.kt [hunk 7]
```diff
@@ -70,13 +70,18 @@ class ProblemsRequests {
...
             val contests = store.state.contests.contests
             val mapContests = contests.associateBy { contest -> contest.id }
+            var amountOfBoundedProblems = 0
             problems.forEach { problem ->
-                problem.contestName = mapContests[problem.contestId]?.name
-                problem.contestTime = mapContests[problem.contestId]?.time
+                if (mapContests[problem.contestId] != null) {
+                    problem.contestName = mapContests[problem.contestId]?.name.orEmpty()
+                    problem.contestTime = mapContests[problem.contestId]?.time ?: 0
+                    amountOfBoundedProblems++
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/problems/redux/requests/ProblemsRequests.kt]: `Bind` -> `Bound`

File: app/src/main/java/com/bogdan/codeforceswatcher/features/problems/ProblemsAdapter.kt (hunk 3 above)
Comment (bogdan-evtushenko) [reply app/src/main/java/com/bogdan/codeforceswatcher/features/problems/ProblemsAdapter.kt]: It is needed - else -> warning

File: app/src/main/java/com/bogdan/codeforceswatcher/features/problems/redux/requests/ProblemsRequests.kt (hunk 6 above)
Comment (bogdan-evtushenko) [reply app/src/main/java/com/bogdan/codeforceswatcher/features/problems/redux/requests/ProblemsRequests.kt]: )))

**Pull request merged**